src/search/SearchMessagesScreen.js

```diff
@@ -131,6 +149,35 @@ class SearchMessagesScreenInner extends PureComponent<Props, State> {
     this.handleQuerySubmit(e);
   };
 
+  fetchOlder = async () => {
+    if (!this.props.loading && !this.state.foundOldest && !this.state.isFetching) {
+      this.setState({ isFetching: true });
```

Comment:
I'm also a bit nervous about the flag that's used as a mutex here, the `isFetching`, be a React state property. The thing about those is that `setState` may not update it immediately -- so I'm not sure it would be the case that we can't have a second call to this function see `!this.state.isFetching`, and enter this block and start a second request, while a first call is still in the middle of it too.

That can be fixed by making this use a direct property on the object, just like the existing `this.lastIdSuccess` etc. (We'll still want `this.state.isFetching` for the purpose it already serves, but this `fetchOlder` logic will use a new `this.isFetchingOlder` or whatever instead.)

---

There's a second issue around this mutex logic, which is that the query can change. In this version, if we have an outstanding `fetchOlder` request for a previous version of the query, and it's hanging for a while for whatever reason, then that'll stop us from fetching older messages for a new query until it finally finishes.

So instead, we can remember which query it is that we're fetching older messages for. There may be several of these, but all we really care about is whether we're doing so for the latest query -- so that can be summarized by another `lastIdFoo` property, similar to the three we have. That is, we can say: … Hmm, this gets fairly tricky. I think it should work out, but it'll take some careful looking at the code, and that's also a hint that maybe there's a better way to architect this.

This second issue is lower-priority, because it can't produce wrong results, only delayed results. And it isn't a super likely case to happen. So I'll be OK with merging a version of this PR that just uses a single boolean like `this.isFetchingOlder`, and then trying to work that out later.

Reply:
the delay still happens in the latest revision, i.e. if I search foo1, then switch to foo2, then to foo3, the page renders foo1 then moves to foo2 then to foo3 which can be a little annoying.